agent/agent.go

```diff
@@ -241,15 +241,43 @@ func (a *Agent) Start() error {
 		return fmt.Errorf("Failed to setup node ID: %v", err)
 	}
 
+	// create the config for the rpc server/client
+	consulCfg, err := a.consulConfig()
+	if err != nil {
+		return err
+	}
+
 	// Setup either the client or the server.
 	if c.Server {
-		server, err := a.makeServer()
+		serverUp := make(chan int)
+		consulCfg.ServerUp = func() { close(serverUp) }
+
+		server, err := consul.NewServerLogger(consulCfg, a.logger)
 		if err != nil {
-			return err
+			return fmt.Errorf("Failed to start Consul server: %v", err)
 		}
 
 		a.delegate = server
-		a.state.Init(c, a.logger, server)
+
+		// wait for the server to be ready
+		// then trigger an anti-entropy run ???
+
+		// this needs review since there is no for the state to be
```

Comment:
Traced this back a bit, and this isn't obsolete. The callback will ultimately be triggered by two places:

1. https://github.com/hashicorp/consul/blob/master/agent/consul/serf.go#L105-L111
2. https://github.com/hashicorp/consul/blob/master/agent/consul/client.go#L290-L293

Anti-entropy normally runs on a timer, or immediately if the agent changes state (like a new service is registered). The server up trigger is used to get the cluster to converge if the agents are started up without any servers present, or if there's a leadership change to the servers. Without this, you'd have to wait for one of the other triggers, so that could take a very long time to reconverge, such as if the cluster got a totally new set of Consul servers.

Isn't the coupling already at the agent level by this line:

https://github.com/hashicorp/consul/blob/master/agent/agent.go#L778

In the agent that's just an empty `func()` callback, so nothing else at the agent level knows this is related to the local state.